psydac/linalg/solvers.py

```diff
@@ -104,44 +104,49 @@ class ConjugateGradient(InverseLinearOperator):
     def __init__(self, A, *, x0=None, tol=1e-6, maxiter=1000, verbose=False):
 
         assert isinstance(A, LinearOperator)
-        assert A.domain == A.codomain
-        self._solver = 'cg'
+        assert A.domain.dimension == A.codomain.dimension
+        domain = A.codomain
+        codomain = A.domain
+
+        if x0 is not None:
+            assert isinstance(x0, Vector)
+            assert x0.space is codomain
+        else:
+            x0 = codomain.zeros()
+
         self._A = A
-        self._domain = A.codomain
-        self._codomain = A.domain
-        self._space = A.domain
+        self._domain = domain
+        self._codomain = codomain
         self._x0 = x0
+        self._solver = 'cg'
         self._tol = tol
         self._maxiter = maxiter
         self._verbose = verbose
         self._options = {"x0":self._x0, "tol":self._tol, "maxiter": self._maxiter, "verbose": self._verbose}
         self._check_options(**self._options)
-        self._tmps = {"v":A.domain.zeros(), "r":A.domain.zeros(), "p":A.domain.zeros(), 
-                      "lp":A.domain.zeros(), "lv":A.domain.zeros()}
+        self._tmps = {key: A.domain.zeros() for key in ("v", "r", "p", "lp", "lv")}
         self._info = None
 
     def _check_options(self, **kwargs):
```

Comment:
Looking at the handling of the solvers' options with fresh eyes, it seems to me that we could simplify quite a bit of code in this module.

Why do we need the local variables `self._x0`, `self._tol`, `self._maxiter`, and `self._verbose` if we already have the `self._options` dictionary?

Accordingly, in `psydac.linalg.basic` I suggest to replace the three functions
```python
    @abstractmethod
    def _update_options(self):
        pass

    def getoptions(self):
        for key, value in self.options.items():
            print(key, ": ", value)

    def setoptions(self, **kwargs):
        self._check_options(**kwargs)
        for key, value in kwargs.items():
            setattr(self, '_'+key, value)
        self._update_options()
```
with these two:
```python
    def get_options(self):
        return self._options.copy()

    def set_options(self, **kwargs):
        self._check_options(**kwargs)
        self._options.update(kwargs)
```

Reply:
That makes perfect sense, will do :)

Reply:
7ff1d83af9860911a63e7812a43c909dc8033484